Source/DafnyLanguageServer/Workspace/TextDocumentLoader.cs

```diff
@@ -184,126 +206,103 @@ bool loadCanceled
 
-      foreach (var implementationTask in implementationTasks) {
-        cancellationToken.Register(implementationTask.Cancel);
-        try {
-          implementationTask.Run();
-        } catch (InvalidOperationException) {
-          // Thrown in case the task was already cancelled. Requires a Boogie fix to remove.
-        }
-        if (VerifierOptions.GutterStatus) {
-          progressReporter.ReportStartVerifyImplementation(implementationTask.Implementation);
-        }
-      }
+      var result = implementationTasks.Select(task => Verify(document, task, cancellationToken)).Merge().
+        Where(d => implementationTasks.All(t => {
+          var taskStatus = d.ImplementationViewsView![GetImplementationId(t.Implementation)].Status;
+          return taskStatus != PublishedVerificationStatus.Stale;
+        })).Replay();
+      result.Connect();
 
       if (VerifierOptions.GutterStatus) {
-        ReportRealtimeDiagnostics(document, result, progressReporter, cancellationToken);
+        ReportRealtimeDiagnostics(document, result, cancellationToken);
       }
 
       var _ = NotifyStatusAsync(document.TextDocumentItem, result.DefaultIfEmpty(document), cancellationToken);
       return result;
     }
 
-    private IObservable<DafnyDocument> GetVerifiedDafnyDocuments(DafnyDocument document, IReadOnlyList<IImplementationTask> implementationTasks,
-      VerificationProgressReporter progressReporter, CancellationToken cancellationToken) {
+    public IObservable<DafnyDocument> Verify(DafnyDocument dafnyDocument, IImplementationTask implementationTask, CancellationToken cancellationToken) {
 
-      var implementationViews = GetExistingViews(document, implementationTasks);
-      var counterExamples = new ConcurrentStack<Counterexample>();
+      if (VerifierOptions.GutterStatus) {
+        dafnyDocument.GutterProgressReporter!.ReportStartVerifyImplementation(implementationTask.Implementation);
+      }
 
-      var implementationsUpdates = implementationTasks.Select(implementationTask => {
-        var subject = WrapObservable(cancellationToken, implementationTask);
-        return subject.SelectMany(boogieStatus =>
-          HandleStatusUpdate(implementationTask, boogieStatus).ToObservable());
-      });
-      var result = implementationsUpdates.Merge().Replay();
-      result.Connect();
-      return result;
+      try {
+        var observable = implementationTask.Run();
+        cancellationToken.Register(() => {
+          try {
+            implementationTask.Cancel();
+          } catch (InvalidOperationException e) {
+            if (!e.Message.Contains("no ongoing run")) {
+              throw;
+            }
+          }
+        });
+        return GetVerifiedDafnyDocuments(dafnyDocument, implementationTask, observable);
+      } catch (InvalidOperationException e) {
+        if (e.Message.Contains("already completed") || e.Message.Contains("ongoing run")) {
```

Comment:
`"no ongoing run".Contains("ongoing run")` is true... is that a problem?

Could you at least put these strings into constants and define these constant in an appropriate place?

Reply:
> I'll address this with an updated Boogie API so no try/catch is needed: https://github.com/boogie-org/boogie/pull/592

Well, that's more than I asked for, but thanks :-)